src/interchange/src/json.rs

```diff
@@ -8,10 +8,203 @@
 // by the Apache License, Version 2.0.
 
 use std::collections::HashSet;
+use std::fmt;
 
+use itertools::Itertools;
+
+use repr::adt::jsonb::JsonbRef;
 use repr::adt::numeric::{NUMERIC_AGG_MAX_PRECISION, NUMERIC_DATUM_MAX_PRECISION};
-use repr::{ColumnName, ColumnType, ScalarType};
-use serde_json::json;
+use repr::{ColumnName, ColumnType, Datum, RelationDesc, ScalarType};
+use serde_json::{json, Map};
+
+use crate::encode::{column_names_and_types, Encode, TypedDatum};
+
+// Manages encoding of JSON-encoded bytes
+pub struct JsonEncoder {
+    key_columns: Option<Vec<(ColumnName, ColumnType)>>,
+    value_columns: Vec<(ColumnName, ColumnType)>,
+}
+
+impl JsonEncoder {
+    pub fn new(key_desc: Option<RelationDesc>, value_desc: RelationDesc) -> Self {
+        JsonEncoder {
+            key_columns: if let Some(desc) = key_desc {
+                Some(column_names_and_types(desc))
+            } else {
+                None
+            },
+            value_columns: column_names_and_types(value_desc),
+        }
+    }
+
+    pub fn encode_row(&self, row: repr::Row, names_types: &[(ColumnName, ColumnType)]) -> Vec<u8> {
+        let value = encode_datums_as_json(row.iter(), names_types);
+        value.to_string().into_bytes()
+    }
+}
+
+impl Encode for JsonEncoder {
+    fn get_format_name(&self) -> &str {
+        "json"
+    }
+
+    fn encode_key_unchecked(&self, row: repr::Row) -> Vec<u8> {
+        self.encode_row(
+            row,
+            self.key_columns.as_ref().expect("key schema must exist"),
+        )
+    }
+
+    fn encode_value_unchecked(&self, row: repr::Row) -> Vec<u8> {
+        self.encode_row(row, &self.value_columns)
+    }
+}
+
+impl fmt::Debug for JsonEncoder {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("JsonEncoder")
+            .field(
+                "schema",
+                &format!("{:?}", build_row_schema_json(&self.value_columns, "schema")),
+            )
+            .finish()
+    }
+}
+
+/// Encodes a sequence of `Datum` as JSON, using supplied column names and types.
+pub fn encode_datums_as_json<'a, I>(
+    datums: I,
+    names_types: &[(ColumnName, ColumnType)],
+) -> serde_json::value::Value
+where
+    I: IntoIterator<Item = Datum<'a>>,
+{
+    let mut name_idx = 0;
+    let value_fields = names_types
+        .iter()
+        .zip_eq(datums)
+        .map(|((name, typ), datum)| {
+            let name = name.as_str().to_owned();
+            (
+                name,
+                TypedDatum::new(datum, typ.clone()).json(&mut move || {
+                    let ret = format!("record{}", name_idx);
```

Comment:
Is this how it's typically done or a new Materialize thing?

Reply:
I saw this is what we did for Avro records without a name, so I stole it for unnamed JSON records. I can dig around and see if there's a better alternative!